golang/cmd/event-ingest/main.go

```diff
@@ -0,0 +1,135 @@
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+)
+
+const batchSize = 1000
+
+type Event struct {
+	OrganizationID int                    `json:"organization_id,omitempty"`
+	CustomerID     *int                   `json:"customer_id,omitempty"`
+	CustID         string                 `json:"cust_id,omitempty"`
+	EventName      string                 `json:"event_name,omitempty"`
+	TimeCreated    time.Time              `json:"time_created,omitempty"`
+	Properties     map[string]interface{} `json:"properties,omitempty"`
+	IdempotencyID  string                 `json:"idempotency_id,omitempty"`
+	InsertedAt     time.Time              `json:"inserted_at,omitempty"`
+}
+
+func main() {
+	seeds := []string{"localhost:9092"}
+	ctx := context.Background()
+
+	// Setup kafka consumer
+	cl, err := kgo.NewClient(
+		kgo.SeedBrokers(seeds...),
+		kgo.ConsumerGroup("default"),
+		kgo.ConsumeTopics("test-topic"),
+		kgo.DisableAutoCommit(),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer cl.Close()
+
+	//setup db connection
+	db, err := sql.Open("postgres", "postgres://user:password@localhost/dbname?sslmode=disable")
+	if err != nil {
+		panic(err)
+	}
+	defer db.Close()
+
+	// Set up prepared statement for insert
+	stmt, err := db.Prepare(`INSERT INTO events (organization_id, customer_id, cust_id, event_name, time_created, properties, idempotency_id, inserted_at)
+		VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
+		ON CONFLICT DO NOTHING`)
+	if err != nil {
+		panic(err)
+	}
+
+	for {
+		fetches := cl.PollFetches(ctx)
+		if fetches == nil {
+			continue
+		}
+		if fetches.IsClientClosed() {
+			return
+		}
+		if errs := fetches.Errors(); len(errs) > 0 {
+			// All errors are retried internally when fetching, but non-retriable errors are
+			// returned from polls so that users can notice and take action.
+			panic(fmt.Sprint(errs))
+		}
+
+		events := make([]*Event, 0)
+		fetches.EachRecord(func(r *kgo.Record) {
+			event := &Event{}
+			err := json.Unmarshal(r.Value, event)
+			if err != nil {
+				fmt.Printf("Error unmarshalling event: %s\n", err)
+				return
+			}
+			event.CustomerID = nil
+			event.InsertedAt = time.Now()
+			events = append(events, event)
+
+			if len(events) >= batchSize {
+				err = insertEvents(stmt, events, db, ctx)
+				if err != nil {
+					fmt.Printf("Error inserting events into database: %s\n", err)
+				}
+				events = make([]*Event, 0)
+				if err := cl.CommitUncommittedOffsets(context.Background()); err != nil {
+					fmt.Printf("commit records failed: %v", err)
+				}
+			}
+		})
+
+		if len(events) > 0 {
+			err := insertEvents(stmt, events, db, ctx)
+			if err != nil {
+				fmt.Printf("Error inserting events into database: %s\n", err)
+				return
+			}
+			if err := cl.CommitUncommittedOffsets(context.Background()); err != nil {
+				fmt.Printf("commit records failed: %v", err)
+			}
+		}
+	}
+}
+
+func insertEvents(stmt *sql.Stmt, events []*Event, db *sql.DB, ctx context.Context) error {
```

Comment:
As noted above, batching the records is unnecessary. You can do it but I don't think it adds anything, and probably uses more memory. Also, I don't see how the code works, but I'm not super confident since I guess you have actually run it?

In terms of batching, if you want to be a bit fancy, you could encapsulate the batching code in a type, like this, which would hide the actual mechanism away from the main loop, and let you change the algo later. This isn't so much a recommendation as just a suggestion in case you want to try a few different things:

```go
type batch struct {
    tx db.Tx;
    insertStmt db.Stmt;
    count int;
}

func (b *batch) addRecord(r *record) err {
    err := insertStmt.Exec(r.a, r.b, ...)
    if err != nil {
        return err
    }

    count++
    if count < 1000 {
      return nil;
    }

     return tx.Commit()
}
```